### PR TITLE
Support configuring processors to skip or modify traces

### PR DESCRIPTION
**Pull request by timrogers**
This change allows you to configure your tracer with a series of "processors" which can alter traces (for example redacting sensitive data that you don't want to send to the Datadog Agent) or even skip them entirely (for example if you only want to keep "interesting" traces, e.g. ones that were particularly slow).

A processor is a function which is passed a trace, and is expected to either return a trace (which may have been changed from what was passed in) or `undefined` (if you want to skip the trace entirely).

After a trace is formatted and before it is encoded, we will run it through your pipeline of processors until we reach a final result. We will stop part-way if any of your processors
returns `undefined`, in which case we will not append the trace to the queue and it will not be sent to the Agent.

This is inspired by functionality found in the Ruby and Python Datadog APM libraries.

An example, taken from the documentation, is probably useful for showing off this feature:

```javascript
// You can return a modified trace, for example redacting a sensitive URL
const redactPasswords = (trace) => {
  return trace.map((span) => {
    // If the span's resource is /super-secret-endpoint, change it /redacted
    if (span.resource === '/super-secret-endpoint') {
      return Object.assign(span, { resource: '/redacted' })
    else {
      return span
    }
  })
}

// ...or you could even skip certain traces entirely
const onlyKeepSlowTraces = (trace) => {
  // Only keep a trace if one of the spans was slow, taking more than 2 seconds
  if (trace.some((span) => span.duration >= 2 * 1e6))
    return trace
  } else {
    return
  }
}

const tracer = require('dd-trace').init({
  // If a trace is filtered out by the first processor (i.e. it returns `undefined`),
  // subsequent processors will not be invoked
  processors: [onlyKeepSlowTraces, redactPasswords]
})
```

Addresses #271.

### CODE REVIEW

File: src/writer.js [hunk 1]
```diff
@@ -23,53 +24,71 @@ class Writer {
     if (trace.started.length === trace.finished.length) {
       const formattedTrace = trace.finished.map(format)
 
-      log.debug(() => `Encoding trace: ${JSON.stringify(formattedTrace)}`)
-
-      const buffer = encode(formattedTrace)
```
Comment (timrogers) [inline src/writer.js]: This change so that we add the **encoded** trace to the buffer, rather than just the **formatted**, could be avoided if there's a good reason why this approach is better. But then we'd need to change the semantics away from `beforeFlush`, which I picked for consistency with the Ruby implementation of this feature.

File: src/writer.js [hunk 2]
```diff
@@ -23,53 +24,71 @@ class Writer {
     if (trace.started.length === trace.finished.length) {
       const formattedTrace = trace.finished.map(format)
 
-      log.debug(() => `Encoding trace: ${JSON.stringify(formattedTrace)}`)
```
Comment (timrogers) [inline src/writer.js]: We stop logging the encoded trace. We could still do this below in `flush()` if we think it's valuable.

File: src/writer.js [hunk 3]
```diff
@@ -23,53 +24,71 @@ class Writer {
...
-          'Datadog-Meta-Lang': platform.name(),
-          'Datadog-Meta-Lang-Version': platform.version(),
-          'Datadog-Meta-Lang-Interpreter': platform.engine(),
-          'Datadog-Meta-Tracer-Version': tracerVersion,
-          'X-Datadog-Trace-Count': String(this._queue.length)
+      // Take the queue of traces, pass them through any `beforeFlush` callback
+      // and then remove empty elements (where beforeFlush has "skipped" the trace)
+      const traces = this._queue
+        .map(this.beforeFlush.bind(this))
+        .filter((trace) => trace)
```
Comment (timrogers) [inline src/writer.js]: If `beforeFlush` returns a trace that is falsy, we'll skip it

File: src/writer.js (hunk 1 above)
Comment (rochdev) [reply src/writer.js]: There were 2 main reasons for this:

* Encoding is the most expensive operation of the tracer. By encoding it early, the load is distributed. This avoids locking given the asynchronous nature of Node.
* We will eventually implement a check on the total buffer size to flush right away if it gets to large. Since this should be triggered whenever adding a trace, we have to encode it to know the buffer size.

I think what we do in [Python](http://pypi.datadoghq.com/trace/docs/advanced_usage.html#trace-filtering) is closer to what would work in Node.

Basically there could be a `processors` option that takes either an array or a single element. Each element could be either a class with a `process` method, or a function with the same signature as the method. We can revisit the wording later but basically that would be the idea. The processor(s) would then be called on every append whenever the trace is finished.

**Comment by timrogers** [comment]
@rochdev Your feedback makes sense - thanks! If I reworked this to:

* use an array of processors (rather than only supporting one), which are functions which accept a trace and return a trace or undefined
* pass the trace through the processor and then encode the trace, before appending it to the queue

Would that be acceptable?

**Comment by timrogers** [comment]
If so, I should be able to get this reworked nice and quickly ❤️

**Comment by rochdev** [comment]
Definitely! Thanks a lot!

**Comment by timrogers** [comment]
@rochdev Ready for re-review!

File: src/writer.js [hunk 4]
```diff
@@ -67,9 +90,9 @@ class Writer {
     }
   }
 
-  _squeeze (buffer) {
+  _squeeze (trace) {
```
Comment (timrogers) [inline src/writer.js]: I’ll change this back, so please don’t consider it in any review.

**Comment by rochdev** [comment]
The benchmarks are somewhat dependent on the writer internals and are now failing. As soon as that's working I'll do the review :)

**Comment by rochdev** [comment]
Also, please make sure to rebase on v0.6.0 since this is a new feature.

**Comment by timrogers** [comment]
I’ll do that tomorrow. Thanks!

**Comment by timrogers** [comment]
@rochdev Done! This is now rebased on `v0.6.0`, and I've fixed the benchmarks.

**Comment by timrogers** [comment]
I've also switched the target of this PR to be `v0.6.0`.

File: src/writer.js [hunk 5]
```diff
@@ -7,32 +7,55 @@ const encode = require('./encode')
...
   }
 
+  // A trace is considered filtered out if it is undefined, or if it contains no spans
+  isTraceFilteredOut (trace) {
+    return trace === undefined || trace === []
+  }
+
+  applyProcessors (trace) {
+    return this._processors.reduce((currentTrace, processor) => {
+      // If the trace has been filtered out by a previous processor, we don't need
```
Comment (rochdev) [inline src/writer.js]: Prefer explicit code to comments.

File: src/writer.js [hunk 6]
```diff
@@ -7,32 +7,55 @@ const encode = require('./encode')
...
     this._url = url
     this._size = size
+    this._processors = processors
   }
 
   get length () {
     return this._queue.length
   }
 
+  // A trace is considered filtered out if it is undefined, or if it contains no spans
```
Comment (rochdev) [inline src/writer.js]: Prefer explicit code to comments.

File: src/writer.js (hunk 4 above)
Comment (rochdev) [reply src/writer.js]: Yes please :)

File: src/writer.js [hunk 7]
```diff
@@ -7,32 +7,55 @@ const encode = require('./encode')
...
+  isTraceFilteredOut (trace) {
+    return trace === undefined || trace === []
+  }
+
+  applyProcessors (trace) {
```
Comment (rochdev) [inline src/writer.js]: I would probably extract this to an external `Pipeline` class or at the very least an external `process()` function similar to how `format()` and `encode()` are external.

File: src/writer.js [hunk 8]
```diff
@@ -7,32 +7,55 @@ const encode = require('./encode')
...
 
     if (trace.started.length === trace.finished.length) {
-      const formattedTrace = trace.finished.map(format)
+      log.debug(() => `Formatting original trace: ${JSON.stringify(trace.finished)}`)
 
-      log.debug(() => `Encoding trace: ${JSON.stringify(formattedTrace)}`)
+      const formattedTrace = trace.finished.map(format)
 
-      const buffer = encode(formattedTrace)
+      log.debug(() => `Running formatted trace through any configured processors: ${JSON.stringify(formattedTrace)}`)
```
Comment (rochdev) [inline src/writer.js]: I know it's my fault for doing this initially 😅but it would probably make more sense to put these log entries in the formatter/encoder/processor directly.

File: test/writer.spec.js [hunk 9]
```diff
@@ -63,21 +64,87 @@ describe('Writer', () => {
     })
   })
 
+  describe('applyProcessors', () => {
```
Comment (rochdev) [inline test/writer.spec.js]: I'll have to do the test review separately since I've recommended extracting this to an external file.

File: docs/API.md [hunk 10]
```diff
@@ -301,6 +301,7 @@ Options can be configured as a parameter to the [init()](https://datadog.github.
 | flushInterval |                              | 2000      | Interval in milliseconds at which the tracer will submit traces to the agent. |
 | experimental  |                              | {}        | Experimental features can be enabled all at once using boolean `true` or individually using key/value pairs. There are currently no experimental features available. |
 | plugins       |                              | true      | Whether or not to enable automatic instrumentation of external libraries using the built-in plugins. |
+| processors    |                              | []        | Array of functions which can filter out or alter a trace before it is sent to the Agent. See the ["Custom Processors"](#custom-processors) section below for more details. |
```
[the remark this answers is not in the document]
Comment (timrogers) [reply docs/API.md]: How would you see that working? What would we do if someone specified both?

[the remark this answers is not in the document]
Comment (timrogers) [reply docs/API.md]: Perhaps we'd go for, in `config.js`:

```javascript
const configuredProcessors = coalesce(options.processor, options.processors, [])
// ...
this.processors = [].concat(configuredProcessors)
```

**Comment by timrogers** [comment]
@rochdev Sorry that is taking up a fair bit of your time by the way! You guys could probably have done this yourselves in the way you wanted to in way less time 😱

File: src/writer.js [hunk 11]
```diff
@@ -7,32 +7,55 @@ const encode = require('./encode')
...
     return this._queue.length
   }
 
+  // A trace is considered filtered out if it is undefined, or if it contains no spans
+  isTraceFilteredOut (trace) {
+    return trace === undefined || trace === []
+  }
+
+  applyProcessors (trace) {
+    return this._processors.reduce((currentTrace, processor) => {
```
[the remark this answers is not in the document]
Comment (rochdev) [reply src/writer.js]: A valid trace is pretty much an array (of spans). A valid span should have the following structure minimally:

* `trace_id`: UInt64BE instance
* `span_id`: UInt64BE instance
* `name`: String
* `resource`: String
* `service`: String
* `start`: Int64BE instance
* `duration`: Int64BE instance

If `parent_id` is set, it should be a UInt64BE instance
If `error` is set, it should be a Number of `0` or `1`
If `meta` is set, it should be an object containing string:string key/value pairs
If `type` is set, it should be a string

I guess we could say that "format" becomes a pre-processor and there should now be an additional post-processor. In that case we could be a bit more flexible (i.e. allowing other types in certain cases like duration could be a Number and convert it back in the post-processor). This would also change the IDs to use buffers and start/duration to Numbers in the pre-processor, and only convert them to the appropriate type in the post-processor.

File: src/writer.js (hunk 7 above)
Comment (rochdev) [reply src/writer.js]: Nitpick, but I think `run` is probably more semantically correct for a pipeline.

**Comment by rochdev** [comment]
It's actually been very useful to make me think about things I didn't think about at all when I considered how to implement this :)

About trace validation, I put some guidelines in one of the comments. I really think there should be a pre-processor and post-processor at this point. Sorry if this adds a lot of work to the PR 😅

File: src/writer.js [hunk 12]
```diff
@@ -7,32 +7,55 @@ const encode = require('./encode')
...
+    this._processors = processors
   }
 
   get length () {
     return this._queue.length
   }
 
+  // A trace is considered filtered out if it is undefined, or if it contains no spans
+  isTraceFilteredOut (trace) {
+    return trace === undefined || trace === []
```
[the remark this answers is not in the document]
Comment (rochdev) [reply src/writer.js]: The thing is otherwise what happens when you return null?

**Comment by timrogers** [comment]
This has got to be quite a big job now 🙈 But I’ll try to pick it up!

Do you think we’d run the “format” processor first, then any user-defined
processors, then a “validate” processor, then an “encode” processor? Or
would you allow users to define pre- and post-processors?

I would be up for doing some of the transformation at the end after the
processors (e.g. converting numbers to the special types) but perhaps we
could leave that to a future change? It would be making the API more
liberal rather than tightening it, so it wouldn’t be a breaking change.

**Comment by rochdev** [comment]
Well basically both the pre and post processors would be variations of the current formatter. The pre-processor would be a simplified version (i.e. no type conversions for int64), and the post-processor would simply ensure the correct structure and drop anything that is completely invalid.

I'd rather have this right away since otherwise it means that processors will need to handle for example  Int64BE instances which will no longer be what they receive in the future.

It would even be better to have a typed object to make all of this clearer, but that can be done at a later time.

**Comment by timrogers** [comment]
Great - I’ll proceed on that basis.

**Comment by timrogers** [comment]
I've done some more work on this this morning, and am not far off being done, but will need to continue another time.

**Comment by timrogers** [comment]
I'm not going to have time to work on this, so I'm going to close it so it doesn't appear active. Sorry!

**Comment by rochdev** [comment]
Feel free to keep it opened, I can continue the work directly from your PR when we get to implementing this.

**Comment by LukasAV** [comment]
Hi, I was also looking for the possibility to filter sensitive data out of traces in Node.js. Is there by now an other way to achieve this? Or are there plans to continue this implementation?

**Comment by rochdev** [comment]
@LukasAV There are a few constructs that can be used to filter sensitive data such as span hooks and a few configuration options on the trace agent. What data exactly are you trying to filter out?

**Comment by LukasAV** [comment]
Thanks for your quick reply. I need to filter out a certain query parameter with sensitive value from the trace URL for all traces. Eg.

`http://url.com?a=keep&sensitive=REMOVED`

**Comment by rochdev** [comment]
@LukasAV In that case you can use [span hooks](https://datadoghq.dev/dd-trace-js/#span-hooks) to alter the metadata of the request span right before it's finished.

**Comment by LukasAV** [comment]
Thank you very much. I got it to work and also discovered that with the latest version all query parameters are removed by default for the http plugin (they were not removed in 0.17.1). From the code I could not see an option to enable these again or is there one? Otherwise I have to add all query params I want to keep also using span hooks.

**Comment by rochdev** [comment]
@LukasAV Generally speaking I would recommend adding any query parameter that may be useful to you as their own tags instead. This will make it easier to search and analyze based on the specific query parameter value instead of the entire URL which would contain the combination of all params. That can also be done in a span hook to add these tags specifically. Would that work for your use case?

**Comment by LukasAV** [comment]
Yes, that would work for me. Thanks for your help.
